File: 1080-insufficient-nodes-in-root-to-leaf-paths/1080-insufficient-nodes-in-root-to-leaf-paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    TreeNode* function(TreeNode* root, int &limit, int sum){
        if(root == NULL) return NULL;
        //main condition
        if(root->left == NULL && root->right == NULL){
            if(sum+root->val < limit) return NULL;
            else return root;
        }
        
        TreeNode* l = function(root->left, limit, sum+root->val);
        TreeNode* r = function(root->right, limit, sum+root->val);
        
        if(l == NULL) root->left = NULL;
        if(r == NULL) root->right = NULL;
        if(l == NULL && r == NULL) return NULL;
        
        return root;
    }
public:
    TreeNode* sufficientSubset(TreeNode* root, int limit) {
        int sum = 0;
        //Use preorder Travesal
        return function(root, limit, sum);
    }
};

File: 1080-insufficient-nodes-in-root-to-leaf-paths/1080-insufficient-nodes-in-root-to-leaf-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sufficientSubset(TreeNode root, int limit) {
        int sum = 0;
        return function(root, limit, sum);
    }
    
    TreeNode function(TreeNode root, int limit, int sum){
        if(root == null) return null;
        if(root.left == null && root.right == null){
            if(sum+root.val < limit) return null;
            else return root;
        }
        //call left
        TreeNode l = function(root.left, limit, sum+root.val);
        //call right
        TreeNode r = function(root.right, limit, sum+root.val);
        
        if(l == null) root.left = null;
        if(r == null) root.right = null;
        if(l == null && r == null) return null;
        return root;
    }
}